pkg/semtech-loramac/contrib/semtech_loramac.c

```diff
@@ -0,0 +1,631 @@
+/*
+ * Copyright (C) 2017 Inria
+ *               2017 Inria Chile
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ * @ingroup     pkg_semtech-loramac
+ *
+ * @file
+ * @brief       Implementation of public API for Semtech LoRaMAC
+ *
+ *              This implementation is an adaption of the applications provided
+ *              on the Semtech Lora-net repository.
+ *
+ *              The LoRaMAC stack and the SX127x driver run in there own thread
+ *              and IPC messages are exchanged to control the MAC.
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ * @author      Jose Alamos <jose.alamos@inria.cl>
+ * @}
+ */
+
+#include <string.h>
+
+#include "msg.h"
+#include "shell.h"
+
+#include "semtech_loramac.h"
+
+#include "sx127x.h"
+#include "sx127x_params.h"
+#include "sx127x_netdev.h"
+
+#include "net/netdev.h"
+#include "net/loramac.h"
+
+#include "semtech-loramac/board.h"
+#include "LoRaMac.h"
+#include "region/Region.h"
+#include "utilities.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define LORAMAC_MSG_QUEUE                           (16U)
+#define LORAMAC_STACKSIZE                           (THREAD_STACKSIZE_DEFAULT)
+
+#define LORAWAN_MAX_JOIN_RETRIES                    (3U)
+
+#if defined(REGION_EU868)
+#define LORAWAN_DUTYCYCLE_ON                        (true)
+#define USE_SEMTECH_DEFAULT_CHANNEL_LINEUP          (1)
+
+#if (USE_SEMTECH_DEFAULT_CHANNEL_LINEUP)
+#define LC4                { 867100000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC5                { 867300000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC6                { 867500000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC7                { 867700000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC8                { 867900000, 0, { ( ( DR_5 << 4 ) | DR_0 ) }, 0 }
+#define LC9                { 868800000, 0, { ( ( DR_7 << 4 ) | DR_7 ) }, 2 }
+#define LC10               { 868300000, 0, { ( ( DR_6 << 4 ) | DR_6 ) }, 1 }
+#endif /* USE_SEMTECH_DEFAULT_CHANNEL_LINEUP */
+#endif /* REGION_EU868 */
+
+#if defined(REGION_CN779) || defined(REGION_EU868) || \
+    defined(REGION_IN865) || defined(REGION_KR920)
+#define LORAWAN_APP_DATA_SIZE                       (16U)
+#elif defined(REGION_AS923) || defined(REGION_AU915) || \
+      defined(REGION_US915) || defined(REGION_US915_HYBRID)
+#define LORAWAN_APP_DATA_SIZE                       (11U)
+#else
+#error "Please define a region in the compiler options."
+#endif
+
+static char stack[LORAMAC_STACKSIZE];
+kernel_pid_t _semtech_loramac_pid;
```

Comment:
The next ten lines: can any of the variables be static? Or namespaced?

Reply:
Maybe the keys could be made static in a header file (I'm not sure) since they need to be reachable from different c files. For the moment, `extern` is used.
For `radio_events`, I don't think this is easily doable, because that would require including LoRamac.h from a header file, and this should avoided (otherwise there will be name clashes).

Reply:
Is it fine, if I go for namespaces ?

Reply:
Yes please. All global variables have to be namespaced. All file scope (static) variables *should* start with underscore followed by non-capital letter.